man/changever.1
.TH changever "1" "December 16, 2023" "" "changever manual"

.SH NAME
changever - helps with generating changelog and publishing Node.js package

.SH SYNOPSIS
.B changever
[-c|--config
.I name
] [-l|--log
.I
file
] [-h|--heading
.I
level
]
.br
[-t|--tag-prefix
.I
prefix
] [-l|--logged-types
.I
types
] [-f|--from
.I
hash
]
.br
[-t|--to
.I
hash
] [-u|--try-unshallow] [-p|--path
.I
path
]
.br
[-r|--repo-url
.I
url
] [-o|--override-version
.I
version
]
.br
[-e|--write-changes
.I
file
] [-w|--write-version
.I
file
]
.br
[-a|--assume-patch] [-0|--bump-major-0] [-d|--dry-run]
.br
[-N|--no-failure] [-i|--print-last] [-q|--quiet] [-v|--verbose]
.br
[-V|--version] [-h|--help]
.I command
[
.I parameters
]
.RE

.SH DESCRIPTION
changever is changelog-driven version manager, which helps with generating
changelog and publishing a new version of a Node.js package.

.SH COMMANDS
.B
.IP "init"
generate a config file with defaults
.B
.IP "version"
prepare the current module for publishing a new version
(update changelog, bump version, commit and tag the change)
.B
.IP "publish"
publish a new version prepared earlier by `changever version`
(push the change and tag, publish package, create a github release)
.B
.IP "release"
perform both `changever version` and `changever publish`

.SH OPTIONS AND PARAMETERS FOR LIST
.B
.IP "-g|--global"
inspect contents of ~./vmodules (default is ./[src/]modules)
.B
.IP "[pkg names]"
names of the package to print its version

.SH PARAMETERS FOR LINK AND UNLINK
.B
.IP "[pkg names]"
name of the package if guessing is not reliable

.SH PARAMETERS FOR VERSION AND PUBLISH
.B
.IP "[version]"
version if the changelog update is disabled
(also major, minor or patch for bumping the existing version)

.SH OPTIONS FOR VERSION, PUBLISH AND RELEASE
.B
.IP "--no-changes"
do not update the changelog
.B
.IP "--no-bump"
do not bump the version in the package manifest
.B
.IP "--no-commit"
do not commit the changes during publishing
.B
.IP "--no-tag"
do not tag the commit during publishing
.B
.IP "--no-push"
do not push the commit and tag during publishing
.B
.IP "--no-release"
do not create a new github release
.B
.IP "--no-failure"
do not fail in case of no version change or release
.B
.IP "--nc-args args"
extra arguments for newchanges, enclosed in quotes
.B
.IP "-y|--yes"
answer the push and reelase confirmations with "yes"
.B
.IP "-d|--dry-run"
only print what would be done without doing it
.B
.IP "-v|--verbose"
print the new changes on the console too

.SH COMMON OPTIONS
.B
.IP "-c|--config name"
file name or path of the config file
.B
.IP "-V|--version"
print the version of the executable and exit
.B
.IP "-h|--help"
print the usage information and exit

.RE

.SH EXAMPLES
.B
.IP "changever version"
Make changes to bump a package version using git log and commit them.
.B
.IP "changever publish -vd"
Print what would happen if the committed changes were pushed and published.

.SH BUGS
changever's bug tracker is located at <https://github.com/prantlf/changever/issues>

.SH DOCUMENTATION
See the website for documentation <https://github.com/prantlf/changever#readme>

.SH AUTHOR
Currently maintained by Ferdinand Prantl <prantlf@gmail.com>


CHANGELOG.md
# Changes

## [0.3.0](https://github.com/prantlf/changever/compare/v0.2.0...v0.3.0) (2023-12-18)

### Features

* Allow disabling npm publishing ([cd53a8f](https://github.com/prantlf/changever/commit/cd53a8f7bc3bd997b7d49d23e3b22e789a14099c))

## [0.2.0](https://github.com/prantlf/changever/compare/v0.1.0...v0.2.0) (2023-12-17)

### Features

* Propagate dry run to npm ([5d34bf3](https://github.com/prantlf/changever/commit/5d34bf3216c515b361b448150f2c93d698e01592))
* Authenticate npm using process environment ([063c82b](https://github.com/prantlf/changever/commit/063c82bf725e6b5f1734b6087d7d91a15872e4c0))

### Bug Fixes

* Use proper parameter for npm verbosity ([0d1e210](https://github.com/prantlf/changever/commit/0d1e21074fe9fc7baf4cdd2770e35d1dee722a19))

## [0.1.0](https://github.com/prantlf/changever/compare/v0.0.0...v0.1.0) (2023-12-16)

### Features

* Initial release ([6a4f12b](https://github.com/prantlf/changever/commit/6a4f12b624e55b163fac4083237daf147e3c0120))

## 0.0.0 (2023-12-16)

Project creation.


src/changever.v
import prantlf.cli { Cli, Env, run }
import prantlf.debug { new_debug }

const version = '0.3.0'

const usage = 'Changelog-driven version manager - helps with generating changelog and publishing a new version of a Node.js package.

Usage: changever [options] <command> [parameters]

Commands:
  init          generate a config file with defaults
  version       prepare the current module for publishing a new version
                (update changelog, bump version, commit and tag the change)
  publish       publish a new version prepared earlier by `changever version`
                (push the change and tag, publish package, create gh release)
  release       perform both `changever version` and `changever publish`

Parameters for version and publish:
  [<version>]   version if the changelog update is disabled
                (also major, minor or patch for bumping the existing version)

Options for version, publish and release:
  --no-changes       do not update the changelog
  --no-bump          do not bumpt the version in package.json and in the lock
  --no-commit        do not commit the changes during publishing
  --no-tag           do not tag the commit during publishing
  --no-push          do not push the commit and tag during publishing
  --no-publish       do not publish a new npm package version
  --no-release       do not create a new github release
  --no-failure       do not fail in case of no version change or release
	--nc-args <args>   extra arguments for newchanges, enclosed in quotes
  -y|--yes           answer the push and reelase confirmations with "yes"
  -d|--dry-run       only print what would be done without doing it
  -v|--verbose       print the new changes on the console too

Common options:
  -c|--config <name>  file name of path of the config file
  -V|--version        print the version of the executable and exits
  -h|--help           print the usage information and exits

Examples:
  $ changever version
  $ changever publish -vd'

struct Opts {
	changes   bool = true
	bump      bool = true
	commit    ?bool
	tag       ?bool
	push      bool = true
	publish   bool = true
	release   bool = true
	failure   bool = true
	nc_args   string @[json: 'nc-args']
	yes       bool
	dry_run   bool   @[json: 'dry-run']
	verbose   bool
	gh_token  string @[json: 'gh-token']
	npm_token string @[json: 'npm-token']
}

const d = new_debug('vp')

fn main() {
	run(Cli{
		usage: usage
		version: version
		options_anywhere: true
		cfg_opt: 'c'
		cfg_gen_arg: 'init'
		cfg_file: '.changever'
		env: Env.both
	}, body)
}

fn body(mut opts Opts, args []string) ! {
	if args.len == 0 {
		return error('Command is missing.')
	}

	first_arg := if args.len > 1 {
		args[1]
	} else {
		''
	}

	command := args[0]
	match command {
		'version' {
			commit := opts.commit or { true }
			tag := opts.tag or { true }
			create_version(first_arg, commit, tag, &opts)!
		}
		'publish' {
			commit := opts.commit or { false }
			tag := opts.tag or { false }
			publish(commit, tag, &opts)!
		}
		'release' {
			commit := opts.commit or { true }
			tag := opts.tag or { true }
			version_and_publish(first_arg, commit, tag, &opts)!
		}
		else {
			return error('Command "${command}" is invalid.')
		}
	}
}
